### PR TITLE
[mlir][mesh] Add verification and canonicalization for some collectives

### PR DESCRIPTION
**Pull request by sogartar**
Add verification and canonicalization for
broadcast, gather, recv, reduce, scatter, send and shift.

The canonicalizations only remove trivial collectives with empty mesh_axes attrubutes.

### CODE REVIEW

**Comment by sogartar** [comment]
@yaochengji, could you review this PR?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 1]
```diff
@@ -510,35 +537,94 @@ void AllToAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
 
 LogicalResult ReduceOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  // TODO
-  return failure();
+  auto mesh = getMeshAndVerifyAxes(*this, symbolTable);
+  if (failed(mesh)) {
+    return failure();
+  }
+  auto meshShape = mesh.value().canonicalDimSizes();
+  if (failed(verifyInGroupDevice(getLoc(), getRootAttrName(), getRoot(),
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:617]: `getRoot() && failed(verifyInGroupDevice(...))`?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 2]
```diff
@@ -510,35 +537,94 @@ void AllToAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
 LogicalResult
 BroadcastOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  // TODO
-  return failure();
+  auto mesh = getMeshAndVerifyAxes(*this, symbolTable);
+  if (failed(mesh)) {
+    return failure();
+  }
+  auto meshShape = mesh.value().canonicalDimSizes();
+  if (failed(verifyInGroupDevice(getLoc(), getRootAttrName(), getRoot(),
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:545]: `getRoot() && failed(verifyInGroupDevice(...))`?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 3]
```diff
@@ -510,35 +537,94 @@ void AllToAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
 
 LogicalResult GatherOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  // TODO
-  return failure();
+  auto mesh = getMeshAndVerifyAxes(*this, symbolTable);
+  if (failed(mesh)) {
+    return failure();
+  }
+  auto meshShape = mesh.value().canonicalDimSizes();
+  if (failed(verifyInGroupDevice(getLoc(), getRootAttrName(), getRoot(),
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:568]: `getRoot() && failed(verifyInGroupDevice(...))`?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 4]
```diff
@@ -567,26 +653,74 @@ void ReduceScatterOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
 
 LogicalResult ScatterOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  // TODO
-  return failure();
+  auto mesh = getMeshAndVerifyAxes(*this, symbolTable);
+  if (failed(mesh)) {
+    return failure();
+  }
+  auto meshShape = mesh.value().canonicalDimSizes();
+  if (failed(verifyInGroupDevice(getLoc(), getRootAttrName(), getRoot(),
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:661]: `getRoot() && failed(verifyInGroupDevice(...))`?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 5]
```diff
@@ -567,26 +653,74 @@ void ReduceScatterOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
 
 LogicalResult SendOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  // TODO
-  return failure();
+  auto mesh = getMeshAndVerifyAxes(*this, symbolTable);
+  if (failed(mesh)) {
+    return failure();
+  }
+  auto meshShape = mesh.value().canonicalDimSizes();
+  if (failed(verifyInGroupDevice(getLoc(), getDestinationAttrName(),
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:687]: `getDestination() && failed(verifyInGroupDevice(...))`?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 6]
```diff
@@ -567,26 +653,74 @@ void ReduceScatterOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
+                       << ". It must be one of the grouping mesh axes.";
+  }
+
+  return success();
+}
+
+void ShiftOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                          MLIRContext *context) {
+  // TODO: remove op when offset is 0 or if it is a rotate with and
+  // offset % sift_axis_mesh_dim_size == 0.
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:723]: typo: `shift_axis_mesh_dim_size`

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 2 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:545]: There is always a root in broadcast.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 3 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:568]: There is always a root in gather.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 1 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:617]: There is always a root in reduce.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 4 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:661]: There is always a root in scatter.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 5 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:687]: There is always a destination in send.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 3 above)
Comment (yaochengji) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:568]: But it might be a `Value`, not an attr?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 6 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:723]: Fixed it.

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp (hunk 3 above)
Comment (sogartar) [reply mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:568]: E.g. `root = [1, ?, 2]` then `root_dynamic` would have to have 1 value corresponding to dimension 1.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks

**Comment by sogartar** [comment]
Rebased it.

**Pull request merged**